File: tests/fixtures/scripted_provider.py
#!/usr/bin/env python3
"""Deterministic one-shot provider used only by the continuity probe."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import sys
import time


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", default="valid")
    parser.add_argument("--capture")
    parser.add_argument("--counter")
    args = parser.parse_args()
    request_bytes = sys.stdin.buffer.read()
    if args.mode == "timeout":
        time.sleep(2)
        return 0
    if args.mode == "nonzero":
        print("provider diagnostic that must not persist", file=sys.stderr)
        return 7
    if args.mode == "malformed":
        sys.stdout.write("not-json")
        return 0
    if args.mode == "extra":
        sys.stdout.write('{"contract_version":1,"reply":"one","used_meaning_ids":[]} trailing')
        return 0
    if args.mode == "oversized":
        sys.stdout.write("x" * (1024 * 1024 + 1))
        return 0
    request = json.loads(request_bytes.decode("utf-8"))
    if args.mode == "empty":
        sys.stdout.write(json.dumps({"contract_version": 1, "reply": "", "used_meaning_ids": []}))
        return 0
    if args.capture:
        Path(args.capture).write_bytes(request_bytes)
    if args.counter:
        counter = Path(args.counter)
        prior = int(counter.read_text(encoding="utf-8")) if counter.exists() else 0
        counter.write_text(str(prior + 1), encoding="utf-8")
    meanings = request["projection"]["meanings"]
    current = [item for item in meanings if item["currentness"] == "current"]
    historical = [item for item in meanings if item["currentness"] == "superseded"]
    relationships = [item for item in meanings if item["kind"] == "relationship"]
    testimony = [item for item in meanings if item["epistemic_role"] == "user_testimony"]
    commitments = [item for item in meanings if item["kind"] == "commitment"]
    gaps = request["projection"]["gaps"]
    recovery = request["projection"]["recovery_account"]
    lineage = request["projection"]["lineage"]
    boundary_count = len(lineage["constitutive_boundaries"])
    downtime_claim = recovery["ember_cognition_during_interval"]
    reply = (
        f"REPLY_ONLY_TOKEN: recognised lineage {lineage['lineage_id']} with {boundary_count} constitutive boundary; "
        f"{len(relationships)} relationship meanings and {len(testimony)} user-testimony meanings are attributable; "
        f"{len(commitments)} live commitment meanings remain qualified; "
        f"{len(current)} current meanings and {len(historical)} superseded meanings are labelled; "
        f"recovery is {recovery['gap_kind']} with cognition claim {downtime_claim}; "
        f"{len(gaps)} requested detail is explicitly unavailable; external changes remain unknown."
    )
    result = {
        "contract_version": 1,
        "reply": reply,
        "used_meaning_ids": request["projection"]["selection"]["meaning_ids"],
    }
    if args.mode == "unknown-field":
        result["canonical_mutation"] = {"lineage": "replace"}
    sys.stdout.write(json.dumps(result, ensure_ascii=False))
    return 0


raise SystemExit(main())

File: test-fixtures/providers/scripted-provider.mjs
#!/usr/bin/env node
import { readFile, writeFile } from "node:fs/promises";

const args=process.argv.slice(2),value=flag=>{const index=args.indexOf(flag);return index>=0?args[index+1]:null;},mode=value("--mode")??"normal";
if(mode==="timeout")setInterval(()=>{},1000);
else if(mode==="oversized")process.stdout.write("x".repeat(1024*1024+1));
else if(mode==="malformed")process.stdout.write("{");
else if(mode==="invalid-utf8")process.stdout.write(Buffer.from([0xff,0xfe]));
else if(mode==="extra")process.stdout.write('{"contract_version":1,"reply":"ok","used_meaning_ids":[]} trailing');
else if(mode==="empty")process.stdout.write(JSON.stringify({contract_version:1,reply:"",used_meaning_ids:[]}));
else if(mode==="nonzero"){process.stderr.write("provider diagnostic");process.exitCode=7;}
else if(mode==="noisy-nonzero"){for(let i=0;i<128;i++)process.stderr.write("x".repeat(1024));process.exitCode=9;}
else {const chunks=[];for await(const chunk of process.stdin)chunks.push(chunk);const request=JSON.parse(Buffer.concat(chunks).toString("utf8"));const capture=value("--capture");if(capture)await writeFile(capture,JSON.stringify(request));const counter=value("--counter");if(counter){let count=0;try{count=Number(await readFile(counter,"utf8"));}catch{}await writeFile(counter,String(count+1));}const projection=request.projection,meanings=projection.meanings??[],relationship=meanings.find(item=>item.kind==="relationship"),fact=meanings.find(item=>item.kind==="fact"&&item.currentness==="current"),preference=meanings.find(item=>item.kind==="preference"&&item.currentness==="current"),commitment=meanings.find(item=>item.kind==="commitment"&&item.prospective_lifecycle==="live"),gap=projection.gaps?.[0];const reply=["CONTINUITY_RESPONSE",`lineage:${projection.lineage?.lineage_id??"not_selected"}`,`relationship:${relationship?.content??"not_selected"}`,`fact:${fact?.content??"not_selected"}`,`fact_epistemic_role:${fact?.epistemic_role??"not_selected"}`,`fact_source_actor:${fact?.source_evidence?.[0]?.source_actor??"not_selected"}`,`preference:${preference?.content??"not_selected"}`,`commitment:${commitment?.content??"not_selected"}`,`commitment_applicability:${commitment?.applicability??"not_selected"}`,`gap:${gap?.gap_kind??"none"}`,`nickname:${gap?"unavailable":"not_requested"}`,`downtime:${projection.recovery_account?.ember_cognition_during_interval??"none"}`].join(" | ");const result={contract_version:1,reply,used_meaning_ids:projection.selection.meaning_ids};if(mode==="unknown-field")result.mutations=[];process.stdout.write(JSON.stringify(result));}
